src/devices/src/virtio/block/request.rs

```diff
@@ -224,3 +224,202 @@ impl Request {
+    use memory_model::GuestAddress;
+
+    // Writes at address 0x0 the request_type, reserved, sector.
+    fn write_request_header(mem: &GuestMemory, request_type: u32, sector: u64) {
+        let addr = GuestAddress(0);
+
+        mem.write_obj_at_addr::<u32>(request_type, addr).unwrap();
+        mem.write_obj_at_addr::<u64>(sector, addr.checked_add(8).unwrap())
+            .unwrap();
+    }
+
+    #[test]
+    fn test_read_request_header() {
+        let mem = GuestMemory::new(&[(GuestAddress(0), 0x1000)]).unwrap();
+        let addr = GuestAddress(0);
+        let sector = 123_454_321;
+
+        // Test that all supported request types are read correctly from memory.
+        let supported_request_types = vec![
+            VIRTIO_BLK_T_IN,
+            VIRTIO_BLK_T_OUT,
+            VIRTIO_BLK_T_FLUSH,
+            VIRTIO_BLK_T_GET_ID,
+        ];
+
+        for request_type in supported_request_types {
+            write_request_header(&mem, request_type, sector);
+
+            let request_header = RequestHeader::read_from(&mem, addr).unwrap();
+            assert_eq!(request_header.request_type, request_type);
+            assert_eq!(request_header.sector, sector);
+        }
+
+        // Test that trying to read a request header that goes outside of the
+        // memory boundary fails.
+        assert!(RequestHeader::read_from(&mem, GuestAddress(0x1000)).is_err());
+    }
+
+    #[test]
+    fn test_request_type_from() {
+        assert_eq!(RequestType::from(VIRTIO_BLK_T_IN), RequestType::In);
+        assert_eq!(RequestType::from(VIRTIO_BLK_T_OUT), RequestType::Out);
+        assert_eq!(RequestType::from(VIRTIO_BLK_T_FLUSH), RequestType::Flush);
+        assert_eq!(
+            RequestType::from(VIRTIO_BLK_T_GET_ID),
+            RequestType::GetDeviceID
+        );
+        assert_eq!(RequestType::from(42), RequestType::Unsupported(42));
+    }
+
+    #[test]
+    fn test_execute_error_status() {
+        assert_eq!(
+            ExecuteError::BadRequest(Error::InvalidOffset).status(),
+            VIRTIO_BLK_S_IOERR
+        );
+        assert_eq!(
+            ExecuteError::Flush(io::Error::from_raw_os_error(42)).status(),
+            VIRTIO_BLK_S_IOERR
+        );
+        assert_eq!(
+            ExecuteError::Read(GuestMemoryError::NoMemoryRegions).status(),
+            VIRTIO_BLK_S_IOERR
+        );
+        assert_eq!(
+            ExecuteError::Seek(io::Error::from_raw_os_error(42)).status(),
+            VIRTIO_BLK_S_IOERR
+        );
+        assert_eq!(
+            ExecuteError::Write(GuestMemoryError::NoMemoryRegions).status(),
+            VIRTIO_BLK_S_IOERR
+        );
+        assert_eq!(ExecuteError::Unsupported(42).status(), VIRTIO_BLK_S_UNSUPP);
+    }
+
+    #[test]
+    fn test_parse() {
+        let m = &GuestMemory::new(&[(GuestAddress(0), 0x10000)]).unwrap();
+        let vq = VirtQueue::new(GuestAddress(0), &m, 16);
+
+        assert!(vq.end().0 < 0x1000);
+
+        vq.avail.ring[0].set(0);
```

Comment:
To make this code a bit more readable, I suggest defining the indexes for request_type_descriptor, data_descriptor and status_descriptor to their respective values 0, 1, 2.

Reply:
Done.